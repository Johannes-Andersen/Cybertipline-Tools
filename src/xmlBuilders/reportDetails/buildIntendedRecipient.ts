import { XMLBuilder, XMLParser } from 'fast-xml-parser';
import type { IntendedRecipient } from '../../types';
import { buildDeviceId } from '../common/buildDeviceId';
import { buildEstimatedLocation } from '../common/buildEstimatedLocation';
import { buildIpCapture } from '../common/buildIpCapture';
import { buildPerson } from '../common/buildPerson';

const parser = new XMLParser({
  attributeNamePrefix: '@_',
  ignoreAttributes: false,
  allowBooleanAttributes: true,
});

const builder = new XMLBuilder({
  attributeNamePrefix: '@_',
  ignoreAttributes: false,
  suppressBooleanAttributes: false,
});

export const buildIntendedRecipient = (
  intendedRecipient: IntendedRecipient,
  keyName = 'intendedRecipient',
): string => {
  const {
    estimatedLocation,
    deviceId,
    ipCaptureEvent,
    intendedRecipientPerson,
    accountPermanentlyDisabled,
    accountTemporarilyDisabled,
  } = intendedRecipient;

  const deviceIds = deviceId?.map(
    (e) => parser.parse(buildDeviceId(e)).deviceId,
  );
  const ipCaptureEvents = ipCaptureEvent?.map(
    (e) => parser.parse(buildIpCapture(e)).ipCaptureEvent,
  );
  const estimatedLocations = estimatedLocation
    ? parser.parse(buildEstimatedLocation(estimatedLocation)).estimatedLocation
    : undefined;
  const person = intendedRecipientPerson
    ? parser.parse(buildPerson(intendedRecipientPerson)).person
    : undefined;

  return builder.build({
    [keyName]: {
      intendedRecipientPerson: person,
      espIdentifier: intendedRecipient.espIdentifier,
      espService: intendedRecipient.espService,
      screenName: intendedRecipient.screenName,
      displayName: intendedRecipient.displayName,
      profileUrl: intendedRecipient.profileUrl,
      ipCaptureEvent: ipCaptureEvents,
      deviceId: deviceIds,
      priorCTReports: intendedRecipient.priorCTReports,
      groupIdentifier: intendedRecipient.groupIdentifier,
      accountTemporarilyDisabled: accountTemporarilyDisabled
        ? {
            '#text': !!accountTemporarilyDisabled,
            '@_disabledDate':
              typeof accountTemporarilyDisabled === 'object'
                ? accountTemporarilyDisabled?.disabledDate?.toISOString()
                : undefined,
            '@_reenabledDate':
              typeof accountTemporarilyDisabled === 'object'
                ? accountTemporarilyDisabled.reenabledDate?.toISOString()
                : undefined,
          }
        : undefined,
      accountPermanentlyDisabled: accountPermanentlyDisabled
        ? {
            '#text': !!accountPermanentlyDisabled,
            '@_disabledDate':
              typeof accountPermanentlyDisabled === 'object'
                ? accountPermanentlyDisabled?.disabledDate?.toISOString()
                : undefined,
          }
        : undefined,
      estimatedLocation: estimatedLocations,
      additionalInfo: intendedRecipient.additionalInfo,
    },
  });
};
