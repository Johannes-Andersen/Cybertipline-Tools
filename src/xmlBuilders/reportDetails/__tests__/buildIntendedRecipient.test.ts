import { describe, expect, it } from 'vitest';
import {
  Country,
  DeviceEventName,
  DeviceIdType,
  IPEventName,
  State,
} from '../../../types';
import type { IntendedRecipient } from '../../../types';
import { buildIntendedRecipient } from '../buildIntendedRecipient';

describe('buildIntendedRecipient', () => {
  it('should build an intended recipient with all fields', () => {
    const intendedRecipient: IntendedRecipient = {
      intendedRecipientPerson: {
        firstName: 'Thomas',
        lastName: 'Anderson',
        age: 28,
        dateOfBirth: new Date('1996-02-10'),
        phone: [
          {
            number: '555-4444',
            verified: true,
          },
        ],
        email: [
          {
            email: 'thomas.anderson@email.com',
            verified: true,
          },
        ],
        address: [
          {
            address: '456 Tech Street',
            city: 'Chicago',
            state: State.Illinois,
            zipCode: '234',
          },
        ],
      },
      espIdentifier: 'USER654321',
      espService: 'Gaming Platform Alpha',
      screenName: 'neo_matrix',
      displayName: ['Neo', 'The One'],
      profileUrl: ['https://gaming-alpha.com/neo_matrix'],
      ipCaptureEvent: [
        {
          ipAddress: '192.168.4.100',
          eventName: IPEventName.Login,
          dateTime: new Date('2024-01-15T15:30:00Z'),
          possibleProxy: false,
        },
      ],
      deviceId: [
        {
          idType: DeviceIdType.IMEI,
          idValue: '123456789054321',
          eventName: DeviceEventName.Login,
          dateTime: new Date('2024-01-15T15:30:00Z'),
        },
      ],
      priorCTReports: [987654],
      groupIdentifier: 'Matrix Gamers',
      accountTemporarilyDisabled: {
        disabledDate: new Date('2024-01-10T00:00:00Z'),
        reenabledDate: new Date('2024-01-12T00:00:00Z'),
      },
      estimatedLocation: {
        city: 'Chicago',
        region: 'Illinois',
        countryCode: Country.UnitedStates,
        verified: true,
        timestamp: new Date('2024-01-15T16:00:00Z'),
      },
      additionalInfo:
        'User has multiple accounts across different gaming platforms',
    };

    const result = buildIntendedRecipient(intendedRecipient);

    expect(result).toMatchSnapshot();
  });

  it('should build an intended recipient with minimal online presence', () => {
    const intendedRecipient: IntendedRecipient = {
      screenName: 'unknown_gamer',
      ipCaptureEvent: [
        {
          ipAddress: '192.168.4.200',
          eventName: IPEventName.Login,
        },
      ],
      accountPermanentlyDisabled: true,
    };

    const result = buildIntendedRecipient(intendedRecipient);

    expect(result).toMatchSnapshot();
  });

  it('should build an intended recipient with person details and account status', () => {
    const intendedRecipient: IntendedRecipient = {
      intendedRecipientPerson: {
        firstName: 'Mark',
        lastName: 'Williams',
        age: 25,
      },
      screenName: 'mark_gaming',
      accountTemporarilyDisabled: true,
      accountPermanentlyDisabled: {
        disabledDate: new Date('2024-01-15T00:00:00Z'),
      },
      additionalInfo: 'Account disabled due to multiple violations',
    };

    const result = buildIntendedRecipient(intendedRecipient);

    expect(result).toMatchSnapshot();
  });

  it('should build an intended recipient with multiple devices and group info', () => {
    const intendedRecipient: IntendedRecipient = {
      espIdentifier: 'GAMER123',
      espService: 'Multi-Gaming Network',
      screenName: 'pro_gamer',
      deviceId: [
        {
          idType: DeviceIdType.IMEI,
          idValue: '555666777888999',
          eventName: DeviceEventName.Login,
        },
        {
          idType: DeviceIdType.ICCID,
          idValue: '11122233344455566677',
          eventName: DeviceEventName.Registration,
        },
      ],
      groupIdentifier: 'Pro Gamers Alliance',
      priorCTReports: [111222, 333444],
      estimatedLocation: {
        city: 'New York',
        region: 'New York',
        countryCode: Country.UnitedStates,
        verified: false,
      },
    };

    const result = buildIntendedRecipient(intendedRecipient);

    expect(result).toMatchSnapshot();
  });
});
