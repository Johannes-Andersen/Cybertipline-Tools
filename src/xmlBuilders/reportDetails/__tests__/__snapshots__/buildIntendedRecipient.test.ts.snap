// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`buildIntendedRecipient > should build an intended recipient with all fields 1`] = `"<intendedRecipient><intendedRecipientPerson><firstName>Thomas</firstName><lastName>Anderson</lastName><phone verified="true">555-4444</phone><email verified="true">thomas.anderson@email.com</email><address><address>456 Tech Street</address><city>Chicago</city><zipCode>234</zipCode><state>IL</state></address><age>28</age><dateOfBirth>1996-02-10</dateOfBirth></intendedRecipientPerson><espIdentifier>USER654321</espIdentifier><espService>Gaming Platform Alpha</espService><screenName>neo_matrix</screenName><displayName>Neo</displayName><displayName>The One</displayName><profileUrl>https://gaming-alpha.com/neo_matrix</profileUrl><ipCaptureEvent><ipAddress>192.168.4.100</ipAddress><eventName>Login</eventName><dateTime>2024-01-15T15:30:00.000Z</dateTime><possibleProxy>false</possibleProxy></ipCaptureEvent><deviceId><idType>IMEI</idType><idValue>123456789054321</idValue><eventName>Login</eventName><dateTime>2024-01-15T15:30:00.000Z</dateTime></deviceId><priorCTReports>987654</priorCTReports><groupIdentifier>Matrix Gamers</groupIdentifier><accountTemporarilyDisabled disabledDate="2024-01-10T00:00:00.000Z" reenabledDate="2024-01-12T00:00:00.000Z">true</accountTemporarilyDisabled><estimatedLocation verified="true" timestamp="2024-01-15T16:00:00.000Z"><city>Chicago</city><region>Illinois</region><countryCode>US</countryCode></estimatedLocation><additionalInfo>User has multiple accounts across different gaming platforms</additionalInfo></intendedRecipient>"`;

exports[`buildIntendedRecipient > should build an intended recipient with minimal online presence 1`] = `"<intendedRecipient><screenName>unknown_gamer</screenName><ipCaptureEvent><ipAddress>192.168.4.200</ipAddress><eventName>Login</eventName></ipCaptureEvent><accountPermanentlyDisabled>true</accountPermanentlyDisabled></intendedRecipient>"`;

exports[`buildIntendedRecipient > should build an intended recipient with multiple devices and group info 1`] = `"<intendedRecipient><espIdentifier>GAMER123</espIdentifier><espService>Multi-Gaming Network</espService><screenName>pro_gamer</screenName><deviceId><idType>IMEI</idType><idValue>555666777888999</idValue><eventName>Login</eventName></deviceId><deviceId><idType>ICCID</idType><idValue>11122233344455566677</idValue><eventName>Registration</eventName></deviceId><priorCTReports>111222</priorCTReports><priorCTReports>333444</priorCTReports><groupIdentifier>Pro Gamers Alliance</groupIdentifier><estimatedLocation verified="false"><city>New York</city><region>New York</region><countryCode>US</countryCode></estimatedLocation></intendedRecipient>"`;

exports[`buildIntendedRecipient > should build an intended recipient with person details and account status 1`] = `"<intendedRecipient><intendedRecipientPerson><firstName>Mark</firstName><lastName>Williams</lastName><age>25</age></intendedRecipientPerson><screenName>mark_gaming</screenName><accountTemporarilyDisabled>true</accountTemporarilyDisabled><accountPermanentlyDisabled disabledDate="2024-01-15T00:00:00.000Z">true</accountPermanentlyDisabled><additionalInfo>Account disabled due to multiple violations</additionalInfo></intendedRecipient>"`;
